packages/master/master.js

```diff
@@ -690,12 +702,22 @@ class ControlConnection extends BaseConnection {
 	async listInstancesRequestHandler(message) {
 		let list = [];
 		for (let instance of db.instances.values()) {
-			list.push({
+			let push_info = {
 				id: instance.config.get("instance.id"),
 				name: instance.config.get("instance.name"),
 				assigned_slave: instance.config.get("instance.assigned_slave"),
+				assigned_slave_name: null,
+				game_port: instance.config.get("factorio.game_port"),
+				public_address: null,
 				status: instance.status,
-			});
+			};
+			// get assigned slave info
+			if (instance.config.get("instance.assigned_slave") !== null) {
+				let assigned_slave_info = findInstancesAssignedSlave(instance.config.get("instance.assigned_slave"));
```

Comment:
And here we are calling this in a loop - so you get a loop inside a loop lookup, which makes it worse - this is an (O)^2 complexity issue.

Reply:
I think you should be more careful about looking in the mapping here it's not for certain that the slave an instance is assigned to exists just because it's not null.

Also the variable name `assigned_slave_info` does not follow the camelCase convention.

Reply:
click Add instance button and this instance unassigned any slaves, `instance.config.get("instance.assigned_slave")` return `null`, so I use `instance.config.get("instance.assigned_slave") !== null` to judge
![11f09008-0506-4f77-96a2-2a15c25a6414](https://user-images.githubusercontent.com/4542000/103216128-2e974480-490d-11eb-82c5-c140dc96392b.gif)